File: mtgSearch/src/main/java/com/dbottillo/adapters/LifeCounterAdapter.java
package com.dbottillo.adapters;

import android.content.Context;
import android.support.v7.widget.CardView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.BaseAdapter;
import android.widget.Button;
import android.widget.ImageButton;
import android.widget.TextView;

import com.dbottillo.R;
import com.dbottillo.resources.Player;

import java.util.List;

public class LifeCounterAdapter extends BaseAdapter {

    public interface OnLifeCounterListener {
        void onRemovePlayer(int position);

        void onEditPlayer(int position);

        void onLifeCountChange(int position, int value);

        void onPoisonCountChange(int position, int value);
    }

    private List<Player> players;
    private Context context;
    private LayoutInflater inflater;
    private OnLifeCounterListener listener;
    private boolean showPoison;

    public LifeCounterAdapter(Context context, List<Player> players, OnLifeCounterListener listener, boolean showPoison) {
        this.players = players;
        this.context = context;
        this.inflater = LayoutInflater.from(context);
        this.listener = listener;
        this.showPoison = showPoison;
    }

    public void setShowPoison(boolean showPoison) {
        this.showPoison = showPoison;
    }

    @Override
    public int getCount() {
        return players.size();
    }

    @Override
    public Player getItem(int i) {
        return players.get(i);
    }

    @Override
    public long getItemId(int i) {
        return getItem(i).getId();
    }

    public View getView(int position, View convertView, ViewGroup parent) {
        final PlayerHolder holder;
        if (convertView == null) {
            convertView = inflater.inflate(R.layout.row_life_counter, parent, false);
            holder = new PlayerHolder(convertView);
            convertView.setTag(holder);
            convertView.setId(position);
        } else {
            holder = (PlayerHolder) convertView.getTag();
        }

        Player player = getItem(position);
        if (player.getDiceResult() > 0) {
            holder.name.setText(context.getString(R.string.row_life_counter_name, player.getName(), player.getDiceResult()));
        } else {
            holder.name.setText(player.getName());
        }
        holder.life.setText(String.valueOf(player.getLife()));
        holder.poison.setText(String.valueOf(player.getPoisonCount()));

        int color = getColorOfPosition(position);
        holder.card.setCardBackgroundColor(color);

        /*int color = R.color.life_counter_red;
        if (player.getLife() > 12) {
            color = R.color.life_counter_green;
        } else if (player.getLife() > 7) {
            color = R.color.life_counter_middle;
        }
        holder.life.setTextColor(context.getResources().getColor(color));
        color = R.color.life_counter_red;
        if (player.getPoisonCount() > 7) {
            color = R.color.life_counter_green;
        } else if (player.getPoisonCount() > 3) {
            color = R.color.life_counter_middle;
        }
        holder.poison.setTextColor(context.getResources().getColor(color));*/

        holder.edit.setTag(position);
        holder.edit.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View remove) {
                listener.onEditPlayer((Integer) remove.getTag());
            }
        });
        holder.remove.setTag(position);
        holder.remove.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View remove) {
                listener.onRemovePlayer((Integer) remove.getTag());
            }
        });
        holder.lifePlusOne.setTag(position);
        holder.lifePlusOne.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View remove) {
                listener.onLifeCountChange((Integer) remove.getTag(), 1);
            }
        });
        holder.lifeMinusOne.setTag(position);
        holder.lifeMinusOne.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View remove) {
                listener.onLifeCountChange((Integer) remove.getTag(), -1);
            }
        });
        holder.lifePlusFive.setTag(position);
        holder.lifePlusFive.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View remove) {
                listener.onLifeCountChange((Integer) remove.getTag(), 5);
            }
        });
        holder.lifeMinusFive.setTag(position);
        holder.lifeMinusFive.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View remove) {
                listener.onLifeCountChange((Integer) remove.getTag(), -5);
            }
        });
        holder.poisonPlusOne.setTag(position);
        holder.poisonPlusOne.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View remove) {
                listener.onPoisonCountChange((Integer) remove.getTag(), 1);
            }
        });
        holder.poisonMinusOne.setTag(position);
        holder.poisonMinusOne.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View remove) {
                listener.onPoisonCountChange((Integer) remove.getTag(), -1);
            }
        });

        holder.poisonContainer.setVisibility(showPoison ? View.VISIBLE : View.GONE);

        return convertView;
    }

    private int getColorOfPosition(int position) {
        int mod = position % 5;
        if (mod == 0) {
            return context.getResources().getColor(R.color.player_1);
        }
        if (mod == 1){
            return context.getResources().getColor(R.color.player_2);
        }
        if (mod == 2) {
            return context.getResources().getColor(R.color.player_3);
        }
        if (mod == 3) {
            return context.getResources().getColor(R.color.player_4);
        }
        return context.getResources().getColor(R.color.player_5);
    }

    static class PlayerHolder {
        CardView card;
        TextView name;
        TextView life;
        TextView poison;
        View poisonContainer;
        ImageButton edit;
        ImageButton remove;
        Button lifeMinusOne;
        Button lifePlusOne;
        Button lifeMinusFive;
        Button lifePlusFive;
        Button poisonPlusOne;
        Button poisonMinusOne;

        PlayerHolder(View row) {
            card = (CardView) row.findViewById(R.id.life_counter_card);
            name = (TextView) row.findViewById(R.id.player_name);
            life = (TextView) row.findViewById(R.id.player_life);
            poison = (TextView) row.findViewById(R.id.player_poison);
            poisonContainer = row.findViewById(R.id.life_counter_poison_container);
            edit = (ImageButton) row.findViewById(R.id.player_edit);
            remove = (ImageButton) row.findViewById(R.id.player_remove);
            lifePlusOne = (Button) row.findViewById(R.id.btn_life_plus_one);
            lifeMinusOne = (Button) row.findViewById(R.id.btn_life_minus_one);
            lifePlusFive = (Button) row.findViewById(R.id.btn_life_plus_five);
            lifeMinusFive = (Button) row.findViewById(R.id.btn_life_minus_five);
            poisonPlusOne = (Button) row.findViewById(R.id.btn_poison_plus_one);
            poisonMinusOne = (Button) row.findViewById(R.id.btn_poison_minus_one);
        }
    }

}

File: MTGSearch/src/main/java/com/dbottillo/mtgsearchfree/ui/lifecounter/LifeCounterAdapter.kt
package com.dbottillo.mtgsearchfree.ui.lifecounter

import android.content.Context
import android.support.v4.content.ContextCompat
import android.support.v7.widget.CardView
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.ImageButton
import android.widget.TextView
import com.dbottillo.mtgsearchfree.R
import com.dbottillo.mtgsearchfree.model.Player

class LifeCounterAdapter(val players: List<Player>, val listener: OnLifeCounterListener, var showPoisonCounter: Boolean) : RecyclerView.Adapter<RecyclerView.ViewHolder>() {

    companion object {
        private val TYPE_HEADER: Int = 0
        private val TYPE_PLAYER: Int = 1
        private val TYPE_FOOTER: Int = 2
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder? {
        if (viewType == TYPE_HEADER) {
            return HeaderViewHolder(LayoutInflater.from(parent.context).inflate(R.layout.life_counter_header, parent, false))
        }
        if (viewType == TYPE_FOOTER) {
            return FooterViewHolder(LayoutInflater.from(parent.context).inflate(R.layout.life_counter_footer, parent, false))
        }
        return PlayerViewHolder(LayoutInflater.from(parent.context).inflate(R.layout.row_life_counter, parent, false))
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        if (getItemViewType(position) == TYPE_PLAYER) {
            holder as PlayerViewHolder
            val player = players[position - 1]

            if (player.diceResult > 0) {
                holder.name.text = holder.itemView.context.getString(R.string.row_life_counter_name, player.name, player.diceResult)
            } else {
                holder.name.text = player.name
            }
            holder.life.text = player.life.toString()
            holder.poison.text = player.poisonCount.toString()

            val color = getColorOfPosition(holder.itemView.context, position)
            holder.card.setCardBackgroundColor(color)

            holder.edit.setOnClickListener { listener.onEditPlayer(player) }
            holder.remove.setOnClickListener { listener.onRemovePlayer(player) }
            holder.lifePlusOne.setOnClickListener { listener.onLifeCountChange(player, 1) }
            holder.lifeMinusOne.setOnClickListener { listener.onLifeCountChange(player, -1) }
            holder.lifePlusFive.setOnClickListener { listener.onLifeCountChange(player, 5) }
            holder.lifeMinusFive.setOnClickListener { listener.onLifeCountChange(player, -5) }
            holder.poisonPlusOne.setOnClickListener { listener.onPoisonCountChange(player, 1) }
            holder.poisonMinusOne.setOnClickListener { listener.onPoisonCountChange(player, -1) }

            holder.poisonContainer.visibility = if (showPoisonCounter) View.VISIBLE else View.GONE
        }
    }

    override fun getItemCount(): Int {
        return players.size + 2
    }

    override fun getItemViewType(position: Int): Int {
        return when(position){
            0 -> TYPE_HEADER
            players.size + 1 -> TYPE_FOOTER
            else -> TYPE_PLAYER
        }
    }

    class HeaderViewHolder(val row: View) : RecyclerView.ViewHolder(row)

    class FooterViewHolder(val row: View) : RecyclerView.ViewHolder(row)

    class PlayerViewHolder(val row: View) : RecyclerView.ViewHolder(row) {

        val card: CardView = row.findViewById(R.id.life_counter_card) as CardView
        val name: TextView = row.findViewById(R.id.player_name) as TextView
        val life: TextView = row.findViewById(R.id.player_life) as TextView
        val poison: TextView = row.findViewById(R.id.player_poison) as TextView
        val poisonContainer: View = row.findViewById(R.id.life_counter_poison_container)
        val edit: ImageButton = row.findViewById(R.id.player_edit) as ImageButton
        val remove: ImageButton = row.findViewById(R.id.player_remove) as ImageButton
        val lifePlusOne: Button = row.findViewById(R.id.btn_life_plus_one) as Button
        val lifeMinusOne: Button = row.findViewById(R.id.btn_life_minus_one) as Button
        val lifePlusFive: Button = row.findViewById(R.id.btn_life_plus_five) as Button
        val lifeMinusFive: Button = row.findViewById(R.id.btn_life_minus_five) as Button
        val poisonPlusOne: Button = row.findViewById(R.id.btn_poison_plus_one) as Button
        val poisonMinusOne: Button = row.findViewById(R.id.btn_poison_minus_one) as Button
    }

    private fun getColorOfPosition(context: Context, position: Int): Int {
        val mod = position % 5
        if (mod == 0) {
            return ContextCompat.getColor(context, R.color.player_1)
        }
        if (mod == 1) {
            return ContextCompat.getColor(context, R.color.player_2)
        }
        if (mod == 2) {
            return ContextCompat.getColor(context, R.color.player_3)
        }
        if (mod == 3) {
            return ContextCompat.getColor(context, R.color.player_4)
        }
        return ContextCompat.getColor(context, R.color.player_5)
    }

    fun setShowPoison(showPoison: Boolean) {
        showPoisonCounter = showPoison
    }
}
